kartothek/io/dask/repartition.py

```diff
@@ -0,0 +1,92 @@
+
+
+def calculate_repartition_ratio(
+    dataset_factory: DatasetFactory,
+    table_name: str,
+    predicates: Optional[PredicatesType] = None,
+    *,
+    target_nrows_per_partition: int,
+    dataset_metadata_frac: float,
+) -> dd.core.Scalar:
+    """
+    Calculate a `repartition_ratio` according to the dataset input and `target_nrows_per_partition` (the desired
+    number of rows of each dataframe partition).
+
+    Parameters
+    ----------
+    dataset_factory
+        dataset factory
+    table_name
+        Dataset Table name
+    predicates
+        Optional , but predicates to be applied during calculatin repartition ratio
+    target_nrows_per_partition
+        target num of rows per partition
+    dataset_metadata_frac
+        used during collection of dataset metadata
+    Returns
+    -------
+    repartition_ratio
+        calculated repartition_ratio
+    """
+    dataset_metadata = ktk_ddf.collect_dataset_metadata(
+        factory=dataset_factory,
+        predicates=predicates,
+        table_name=table_name,
+        frac=dataset_metadata_frac,
+    )
+    # Once the kartothek issue is resolved, one should use the rowgroup wise
+    # metadata. This would set up future improvements once kartothek is able to
+    # filter RG predicates while collecting statistics (assuming this will ever
+    # be possible this way)
+    # See https://github.com/JDASoftwareGroup/kartothek/pull/352
+    # and https://issues.apache.org/jira/browse/ARROW-10146
+    expected_nrows_per_partition = dataset_metadata[
+        ["partition_label", "number_rows_total"]
+    ].drop_duplicates()
+    mean_nrows_per_partition = expected_nrows_per_partition.number_rows_total.mean()
+    repartition_ratio = target_nrows_per_partition / mean_nrows_per_partition
+    return repartition_ratio
+
+
+def repartition_ddf(ddf: dd.DataFrame, **kwargs) -> dd.DataFrame:
+    """
+    Repartition a dask DataFrame according, dividing the current number of partitions by `repartition_ratio`.
+
+    Parameters
+    ----------
+    ddf
+        dask dataframe
+    repartition_ratio
+        calculated repartition ratio
+    parquet_size
+        deserialized  size of the parquet file.
+    Returns
+    -------
+    ddf
+       dask dataframe after applying repartition on it.
+       We can trigger a dask optimization to fuse the repartitioned tasks properly.
+    """
+    if kwargs["repartition_ratio"]:
+        repartition_ratio = kwargs["repartition_ratio"]
+        with dask.config.set(
+            {"optimization.fuse.ave-width": math.ceil(repartition_ratio)}
+        ):
+            # Minimum amount of partitions is 1.
+            target_npartitions = max(
+                1, int(math.ceil(ddf.npartitions / repartition_ratio))
+            )
+            ddf2 = ddf.repartition(npartitions=target_npartitions)
+            return ddf2
+    elif kwargs["parquet_size"]:
+        ddf2 = ddf.repartition(partition_size=kwargs["parquet_size"])
```

Comment:
I get that we factor the `repartition_ratio` logic out into a separate function, but for `parquet_size` we are just passing the function input directly to `ddf.repartition(partiiton_size=`, I don't see any value this indirection brings atm

Reply:
Not only is this indirection misleading but this also would trigger an entire dataset load. My earlier comment was rather we can achieve this repartitioning by size by *not* calling dask repartition but rather do this ourselves. Please remove this from the current PR since this is a non-trivial extension to the logic and we should do this in a dedicated PR

Reply:
Removed parquet_size related code from here. Added repartition_ratio only.